### PR TITLE
Apply source-build patches

### PR DESCRIPTION
**Pull request by lpatalas**
This PR applies the patches needed to build Arcade using source-build. Related issue: #5719.

Changes:
- Update `SystemNetHttpPackageVersion` from `4.3.3` to `4.3.4` to avoid version downgrade warning when building from source ([Patch 0001](https://github.com/dotnet/source-build/blob/4f0f472e458be20256b478cddcfefb0b6c64a0d5/patches/arcade/0001-Update-SystemNetHttpPackageVersion.patch))
- Do not reference `Microsoft.CodeAnalysis` in `Microsoft.DotNet.Build.Tasks.Packaging.csproj` when building from source because it provides it's own version ([Patch 0002](https://github.com/dotnet/source-build/blob/4f0f472e458be20256b478cddcfefb0b6c64a0d5/patches/arcade/0002-Remove-reference-to-Microsoft-CodeAnalysis-in-Packag.patch))
- Use `System.Text.Json` version set in `$(SystemTextJsonVersion)` instead of hardcoded `4.6.0-preview7.19362.9` ([Patch 0003](https://github.com/dotnet/source-build/blob/4f0f472e458be20256b478cddcfefb0b6c64a0d5/patches/arcade/0003-Update-SystemTextJson-version-from-Package-flow.patch))
- Set `SystemReflectionMetadataVersion` to `1.6.0` and `SystemCollectionsImmutableVersion` to `1.5.0` to avoid version downgrade warnings when building from source. This also required changes in tests because they use `System.Collections.Immutable` to test `HarvestPackage` functionality ([Patch 0004](https://github.com/dotnet/source-build/blob/4f0f472e458be20256b478cddcfefb0b6c64a0d5/patches/arcade/0004-Update-versions-of-some-packages.patch))
- Override `TargetFrameworks` to not use `net472` when building from source ([Patch 0006](https://github.com/dotnet/source-build/blob/4f0f472e458be20256b478cddcfefb0b6c64a0d5/patches/arcade/0006-Don-t-build-for-net472-when-building-from-source.patch))
- Use `$(SystemReflectionMetadataVersion)` instead of hardcoded version `1.6.0` ([Patch 0007](https://github.com/dotnet/source-build/blob/4f0f472e458be20256b478cddcfefb0b6c64a0d5/patches/arcade/0007-Use-SystemReflectionMetadata-from-package-flow.patch))
- Set `ExcludeFromSourceBuild` property to `true` in test projects ([Patch 0008](https://github.com/dotnet/source-build/blob/4f0f472e458be20256b478cddcfefb0b6c64a0d5/patches/arcade/0008-Exclude-Helix-tests-from-source-build.patch))
- Reset `Newtonsoft.Json` version to `9.0.1` in `Directory.Build.props` when building from source ([Patch 0010](https://github.com/dotnet/source-build/blob/4f0f472e458be20256b478cddcfefb0b6c64a0d5/patches/arcade/0010-Update-newtonsoft-json-version-for-arcade.patch))
- Set `MsbuildTaskMicrosoftCodeAnalysisCSharpVersion` to `$(MicrosoftCodeAnalysisCSharpVersion)` when building from source to always use the one built by source-build ([Patch 0011](https://github.com/dotnet/source-build/blob/4f0f472e458be20256b478cddcfefb0b6c64a0d5/patches/arcade/0011-Reassign-MsbuildTaskMicrosoftCodeAnalysisCSharpVersi.patch))
- Extra: Add `Microsoft.DotNet.RemoteExecutor.Tests` project to solution - it seems that it was missing from there for some reason

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -15,4 +15,22 @@
     -->
     <IsShipping>false</IsShipping>
   </PropertyGroup>
+
+  <PropertyGroup Condition="'$(DotNetBuildFromSource)' == 'true'">
+    <!--
+      Source-build builds 2 versions, 9.0.1 and and 12.0.2
+      and the latter is written to Versions.props. Arcade uses
+      the older version, so the Version.props version needs to be updated.
```
Comment (riarenas) [inline Directory.Build.props:23]: I don't understand this comment, especially since versions.props already says 9.01?

Comment (lpatalas) [reply Directory.Build.props]: I reworded the comment in the code to hopefully be a bit more clear:
```
<PropertyGroup Condition="'$(DotNetBuildFromSource)' == 'true'">
  <!--    
    When building using source-build the process is:
    - Newtonsoft.Json versions 9.0.1 and 12.0.2 are built by source-build
    - Version 12.0.2 is written to Version.props
    - Arcade needs to use 9.0.1 so we need to override Version.props value here
  -->
  <NewtonsoftJsonVersion>9.0.1</NewtonsoftJsonVersion>
</PropertyGroup>
```

**Comment by dleeapho** [comment]
@crummel or @dagood PTAL.

File: Directory.Build.props (hunk 1 above)
Comment (riarenas) [reply Directory.Build.props]: Makes sense to me now. Thanks

File: src/Microsoft.DotNet.Build.Tasks.Packaging/tests/Microsoft.DotNet.Build.Tasks.Packaging.Tests.csproj [hunk 2]
```diff
@@ -4,6 +4,7 @@
     <LangVersion>Latest</LangVersion>
     <SignAssembly>false</SignAssembly>
     <NoWarn>xUnit2013</NoWarn>
+    <ExcludeFromSourceBuild>true</ExcludeFromSourceBuild>
```
Comment (dagood) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/tests/Microsoft.DotNet.Build.Tasks.Packaging.Tests.csproj:7]: Shouldn't Arcade detect `.Tests` in `Microsoft.DotNet.Build.Tasks.Packaging.Tests` and automatically exclude it? (Is this patch necessary?)

Same question for other instances.

File: src/Microsoft.DotNet.NuGetRepack/tasks/Microsoft.DotNet.NuGetRepack.Tasks.csproj [hunk 3]
```diff
@@ -2,6 +2,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <TargetFrameworks>net472;netcoreapp2.1</TargetFrameworks>
+    <TargetFrameworks Condition="'$(DotNetBuildFromSource)' == 'true'">netcoreapp2.1</TargetFrameworks>
```
Comment (dagood) [inline src/Microsoft.DotNet.NuGetRepack/tasks/Microsoft.DotNet.NuGetRepack.Tasks.csproj:5]: Would it be more useful (is it possible) to use a property that's defined once and used in each project?

---

Also, we build [microsoft.netframework.referenceassemblies.net472/1.0.0-preview.2](https://github.com/dotnet/source-build-reference-packages/tree/a9cf71cddf11fa52d4e1379a470561da095d31bb/src/targetPacks/ILsrc/microsoft.netframework.referenceassemblies.net472/1.0.0-preview.2) as ref-only assemblies, is this patch necessary anymore? Or am I misunderstanding the purpose of the package (or the patch)?

Comment (crummel) [reply src/Microsoft.DotNet.NuGetRepack/tasks/Microsoft.DotNet.NuGetRepack.Tasks.csproj:5]: Hmm, the net472 patches probably shouldn't be needed anymore, let me try a build.

File: src/Microsoft.DotNet.Build.Tasks.Packaging/tests/Microsoft.DotNet.Build.Tasks.Packaging.Tests.csproj (hunk 2 above)
Comment (dagood) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/tests/Microsoft.DotNet.Build.Tasks.Packaging.Tests.csproj:7]: I'm talking about this auto-exclude logic:

https://github.com/dotnet/arcade/blob/2e89cc293fc511ba4a21b8ee80ae3016e9e446b9/src/Microsoft.DotNet.Arcade.Sdk/tools/Tests.props#L35-L41

File: src/Microsoft.DotNet.NuGetRepack/tasks/Microsoft.DotNet.NuGetRepack.Tasks.csproj (hunk 3 above)
Comment (crummel) [reply src/Microsoft.DotNet.NuGetRepack/tasks/Microsoft.DotNet.NuGetRepack.Tasks.csproj:5]: Looks like an MSBuild or Roslyn issue actually, I think we'll want to keep this for now:
```
/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/src/Microsoft.DotNet.CodeAnalysis/Microsoft.DotNet.CodeAnalysis.csproj : warning NU1603: Microsoft.CodeAnalysis.Common 3.4.1 depends on System.Text.Encoding.CodePages (>= 4.5.1) but System.Text.Encoding.CodePages 4.5.1 was not found. An approximate best match of System.Text.Encoding.CodePages 4.7.0 was resolved. [/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/Arcade.sln]
/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/src/Microsoft.DotNet.GenFacades/Microsoft.DotNet.GenFacades.csproj : warning NU1603: Microsoft.CodeAnalysis.Common 3.4.1 depends on System.Text.Encoding.CodePages (>= 4.5.1) but System.Text.Encoding.CodePages 4.5.1 was not found. An approximate best match of System.Text.Encoding.CodePages 4.7.0 was resolved. [/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/Arcade.sln]
/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/src/Microsoft.DotNet.CoreFxTesting/Microsoft.DotNet.CoreFxTesting.csproj : error NU1202: Package Microsoft.Build 16.4.0 is not compatible with net472 (.NETFramework,Version=v4.7.2). Package Microsoft.Build 16.4.0 supports: netcoreapp2.1 (.NETCoreApp,Version=v2.1) [/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/Arcade.sln]
/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/src/Microsoft.DotNet.Build.Tasks.Configuration/src/Microsoft.DotNet.Build.Tasks.Configuration.csproj : error NU1202: Package Microsoft.Build 16.4.0 is not compatible with net472 (.NETFramework,Version=v4.7.2). Package Microsoft.Build 16.4.0 supports: netcoreapp2.1 (.NETCoreApp,Version=v2.1) [/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/Arcade.sln]
/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/src/Microsoft.DotNet.Arcade.Sdk/Microsoft.DotNet.Arcade.Sdk.csproj : error NU1202: Package Microsoft.Build 16.4.0 is not compatible with net472 (.NETFramework,Version=v4.7.2). Package Microsoft.Build 16.4.0 supports: netcoreapp2.1 (.NETCoreApp,Version=v2.1) [/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/Arcade.sln]
/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/src/Microsoft.DotNet.Build.Tasks.Packaging/src/Microsoft.DotNet.Build.Tasks.Packaging.csproj : error NU1202: Package Microsoft.Build 16.4.0 is not compatible with net472 (.NETFramework,Version=v4.7.2). Package Microsoft.Build 16.4.0 supports: netcoreapp2.1 (.NETCoreApp,Version=v2.1) [/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/Arcade.sln]
/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/src/Microsoft.DotNet.GenFacades/Microsoft.DotNet.GenFacades.csproj : error NU1605: Detected package downgrade: System.Runtime.CompilerServices.Unsafe from 4.7.0 to 4.5.2. Reference the package directly from the project to select a different version.  [/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/Arcade.sln]
/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/src/Microsoft.DotNet.GenFacades/Microsoft.DotNet.GenFacades.csproj : error NU1605:  Microsoft.DotNet.GenFacades -> Microsoft.CodeAnalysis.CSharp 3.4.1 -> Microsoft.CodeAnalysis.Common 3.4.1 -> System.Text.Encoding.CodePages 4.7.0 -> System.Runtime.CompilerServices.Unsafe (>= 4.7.0)  [/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/Arcade.sln]
/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/src/Microsoft.DotNet.GenFacades/Microsoft.DotNet.GenFacades.csproj : error NU1605:  Microsoft.DotNet.GenFacades -> Microsoft.CodeAnalysis.CSharp 3.4.1 -> Microsoft.CodeAnalysis.Common 3.4.1 -> System.Runtime.CompilerServices.Unsafe (>= 4.5.2) [/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/Arcade.sln]
/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/src/Microsoft.DotNet.GenFacades/Microsoft.DotNet.GenFacades.csproj : error NU1202: Package Microsoft.Build 16.4.0 is not compatible with net472 (.NETFramework,Version=v4.7.2). Package Microsoft.Build 16.4.0 supports: netcoreapp2.1 (.NETCoreApp,Version=v2.1) [/home/chris/arcade-no-patch-tb/src/arcade.15f00efd583eab4372b2e9ca25bd80ace5b119ad/Arcade.sln]
```

File: src/Microsoft.DotNet.Build.Tasks.Packaging/tests/Microsoft.DotNet.Build.Tasks.Packaging.Tests.csproj (hunk 2 above)
Comment (lpatalas) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/tests/Microsoft.DotNet.Build.Tasks.Packaging.Tests.csproj:7]: I'll check it

File: src/Microsoft.DotNet.NuGetRepack/tasks/Microsoft.DotNet.NuGetRepack.Tasks.csproj (hunk 3 above)
Comment (dagood) [reply src/Microsoft.DotNet.NuGetRepack/tasks/Microsoft.DotNet.NuGetRepack.Tasks.csproj:5]: Ah yeah, due to upstream patches like https://github.com/dotnet/source-build/blob/master/patches/roslyn/0001-Conditionally-remove-net472-from-TargetFrameworks.patch.

Filed https://github.com/dotnet/source-build/issues/1650 to try to remove these customizations/patches.

File: src/Microsoft.DotNet.Build.Tasks.Packaging/tests/Microsoft.DotNet.Build.Tasks.Packaging.Tests.csproj (hunk 2 above)
Comment (lpatalas) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/tests/Microsoft.DotNet.Build.Tasks.Packaging.Tests.csproj]: This property is indeed applied automatically so I removed it from csproj files. Thanks for this suggestion.

**Comment by lpatalas** [comment]
This PR is good to go but to follow an old tradition of not merging stuff just before your vacation it needs to wait until 20th.

**Pull request merged**